File: zig/src/inline/superscriptRule.zig
const std = @import("std");
const InlineParserState = @import("../types/InlineParserState.zig").InlineParserState;
const InlineRule = @import("../types/InlineRule.zig").InlineRule;
const MarkdownNode = @import("../types/MarkdownNode.zig").MarkdownNode;
const testTagMarks = @import("tagMarksRule.zig").testTagMarks;

pub fn testSuperscript(state: *InlineParserState, parent: *MarkdownNode) bool {
    if (!state.isEscaped and state.i < state.src.len and state.src[state.i] == '^') {
        return testTagMarks("superscript", '^', state, parent, superscriptRule.precedence.?);
    }
    return false;
}

pub const superscriptRule = InlineRule{
    .name = "superscript",
    .@"test" = testSuperscript,
    .precedence = 5,
};

File: web/src/inline/superscriptRule.ts
import type InlineParserState from "../types/InlineParserState";
import type InlineRule from "../types/InlineRule";
import type MarkdownNode from "../types/MarkdownNode";
import { CARET_CODE } from "../utils/charCodes";
import testTagMarks from "./tagMarksRule";

const rule: InlineRule = {
	name: "superscript",
	test: testSuperscript,
	precedence: 5,
};
export default rule;

function testSuperscript(state: InlineParserState, parent: MarkdownNode): boolean {
	if (!state.isEscaped && state.src.charCodeAt(state.i) === CARET_CODE) {
		return testTagMarks(rule.name, "^", state, parent, rule.precedence!);
	}
	return false;
}
